glean-core/src/traits/object.rs

```diff
@@ -0,0 +1,54 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::fmt::Display;
+
+use serde::Deserialize;
+pub use serde::Serialize;
+use serde_json::Value as JsonValue;
+
+/// This type represents all possible errors that can occur when serializing or deserializing an object from/to JSON.
+#[derive(Debug)]
+pub struct ObjectError(serde_json::Error);
+
+impl Display for ObjectError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Display::fmt(&self.0, f)
+    }
+}
+
+impl std::error::Error for ObjectError {}
+
+/// An object that can be serialized into JSON.
+///
+/// Objects are defined by their structure in the metrics definition.
+///
+/// This is essentially a wrapper around serde's `Serialize`/`Deserialize`,
+/// but in a way we can name it for our JSON (de)serialization.
+pub trait ObjectObject {
```

Comment:
I'm not super happy with the naming here, but I can't think of anything more descriptive.
Maybe this should be just `Object`?
It's essentially a hidden name though (only used by the generated code).

Reply:
`trait`s should be verbs (Display, Debug) anyway, right? Right now it's ob'ject ob-ject' which is about disagreeing about an object. (I'm being hilarious)

(We violate this everywhere in the SDK (looking at you, ExtraKeys), so this isn't a hard req. But it might help suggest something nice to us. Like maybe `Objectionable` or `Objectify`?)